Add dma driver for Nios-II MSGDMA core

ramakrishnapallala opened the pull request.

Commit:
ext: hal: altera: Add modular Scatter-Gather DMA HAL driver

Add Altera Nios-II modular Scatter-Gather DMA (mSGDMA) HAL
driver to Zephyr.

The sources are taken from the Altera SDK v17.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
ext: hal: altera: Add wrapper function for Altera HAL runtime API

Add wrapper function for alt_printf() Altera HAL runtime API
to enable Altera HAL drivers on Zephyr.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
arch: nios2: Add _nios2_dcache_flush_no_writeback() routine

Add _nios2_dcache_flush_no_writeback() routine to invalidate
cache contents of address provided.

This routine is ported from Altera Nios-II SDK v17.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
tests: boards: altera_max10: Add test for Nios-II MSGDMA soft IP

Add test for Altera Nios-II MSGDMA soft IP core. This IP on
Altera MAX10 support Memory to Memory DMA transfers.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

rgundi commented on tests/boards/altera_max10/msgdma/src/dma.c:
```diff
@@ -0,0 +1,105 @@
+
+	while (dma_stat == DMA_OP_STAT_NONE) {
+		k_busy_wait(10);
+	}
```
I think using a semaphore/event is a more elegant way to achieve this.

ramakrishnapallala replied:
if I use semaphore or mutex the task is put sleep. My intention of the test is to occupy the CPU while DMA is ongoing.

rgundi commented on arch/nios2/core/cache.c:
```diff
@@ -60,3 +60,33 @@ void _nios2_dcache_flush_all(void)
+	 * multiple of 2 (which it always is).
+	 */
+
+	if (((u32_t) start) & (ALT_CPU_DCACHE_LINE_SIZE - 1)) {
```
I don't understand the purpose of "&"ing start with ALT_CPU_DCACHE_LINE_SIZE - 1. I am not sure i understand what's happening with "unaligned flush request".  May be an example would help?

ramakrishnapallala replied:
it is checking the DWORD alignment and reused from HAL code.

rgundi reviewed (commented):
Code looks good but i am seeking a couple of clarifications. Nothing that prevents this from getting approved though.

ramakrishnapallala commented:
@lpereira @andrewboie Please review this PR.

andrewboie commented on drivers/dma/dma_nios2_msgdma.c:
```diff
@@ -0,0 +1,240 @@
+	 * descriptor construction API's arguments (alt_32 *)
+	 */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wint-conversion"
```
Why do it this way, instead of just casting?

ramakrishnapallala replied:
simple (u32_t *) didn't work, now i have changed it to (alt_u32 *) and it works fine.
I will remove the prama macros. Thanks.

andrewboie commented on drivers/dma/dma_nios2_msgdma.c:
```diff
@@ -0,0 +1,240 @@
+	status = alt_msgdma_standard_descriptor_async_transfer(cfg->msgdma_dev,
+								&cfg->desc);
+	if (status < 0) {
+		SYS_LOG_ERR("DMA transfer error (%d)", status);
```
could move this outside the semaphore take/give section

ramakrishnapallala replied:
Ok

andrewboie commented on drivers/dma/dma_nios2_msgdma.c:
```diff
@@ -0,0 +1,240 @@
+				       ALTERA_MSGDMA_CSR_STOP_MASK);
+
+	status = IORD_ALTERA_MSGDMA_CSR_STATUS(cfg->msgdma_dev->csr_base);
+	if (status & ALTERA_MSGDMA_CSR_STOP_STATE_MASK) {
```
could move this outside the semaphore critical section

ramakrishnapallala replied:
Ok

andrewboie commented on ext/hal/altera/altera_hal/HAL/src/altera_common.c:
```diff
@@ -80,3 +80,7 @@ unsigned int alt_busy_sleep (unsigned int us)
 }
+
+void alt_printf(const char *fmt, ...)
+{
```
what's this function used for? any value in optionally connecting it to printk()?

ramakrishnapallala replied:
Just want to provide a wrapper as it is being used few places in the HAL driver.

For now, I would like to keep this as dummy function, we can see if it needed in future to compliment with printk.

tyrantlizard77 reviewed (commented):
Minor stuff

ramakrishnapallala commented:
@andrewboie, I have addressed the comments and updated the PR. Please have a look.

ramakrishnapallala commented:
@andyross @andrewboie Can you review the nios2 DMA PR?

nashif commented:
you need to resolve those conflicts

ramakrishnapallala commented:
@nashif rebased the PR. Thanks.

Commit:
drivers: dma: Add dma driver for Nios-II MSGDMA core

Add dma driver for Nios-II Modular Scatter-Gather DMA soft IP.

This driver relies upon the Altera HAL msgdma driver for all
the dma core register programming and interrupt handling.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>
